common/models/transducer/loss.py

```diff
@@ -1,18 +1,18 @@
-
 from returnn.tf.util.data import Data
 
 
 def rnnt_loss(source, **_kwargs):
+  """ Computes the RNN-T loss function. Native TF kernel implementation.
+  B: batch, T: time, U:target/labels, V: vocabulary
+  Args:
+      source: function (i: int, as_data: bool = False, ...) -> tf.Tensor|Data, ...)
+            returns one of [output_log_prob, target, base:encoder]
+        output_log_prob (Data): (B, T, U+1, V) log-probabilities
+        real_target (Data): (B, U) -> [V] target labels
+        base:encoder (Data): (B, T, Feat) -> [V] encoder output
+  Returns:
+      cost (tf.Tensor): (B,)
```

Comment:
Instead of specifying the type here, maybe directly in the func signature as annotation.
And then just:
```suggestion
      cost (B,)
```

Reply:
I think I lost it. I am not sure anymore if we want to use signature or not. Isn't it Python3?

Reply:
In the main RETURNN repo (in the main `returnn` package), we still want to support Python 2 (not sure how long, see [here](https://github.com/rwth-i6/returnn/issues/487), but at least now).
For new separate code (e.g. here in `common`), we don't have this restriction. (We don't really have any strict rules at all here so far.)
Even in the main RETURNN repo, the rule is, new optional code does not need to support old versions (Python and TF as well), as long as it would not break existing setups (i.e. it's an optional new feature).

Reply:
However, I also said to not overengineer it. You came up with this `Protocol` thing. I would avoid that. This should be done in RETURNN itself (the definition of such types) (i.e. only once we drop older Python versions).